### ISSUE TITLE
Receiving notifications doesn't work.

### ISSUE THREAD

**Issue by dmitrysync** [ecederstrand/exchangelib#959]
**Describe the bug**
I have tried the 'Synchronization, subscriptions and notifications' from the [documentation](https://ecederstrand.github.io/exchangelib/#synchronization-subscriptions-and-notifications)

This example doesn't work.

``` python
from exchangelib import (
    DELEGATE,
    Account,
    Configuration,
    Credentials,
    NTLM,
    FolderCollection,
)

ACCOUNT_CFG = {
    'exchange': {
        "email": "",
        "password": "",
        "server": '',
    },

def connect(email: str, password: str, server: str = None, autodiscover: bool = False, auth_type=NTLM) -> Account:
    if not server:
        autodiscover = True
        auth_type = None

    credentials = Credentials(username=email, password=password)
    config = Configuration(server=server, credentials=credentials, auth_type=auth_type)

    account = Account(primary_smtp_address=email, config=config, autodiscover=autodiscover, access_type=DELEGATE)
    account.identity.sid = 'S-my-sid'
    account.identity.upn = email

    return account

def subscription(account):
    addr = "**********.ngrok.io"
    callback_url = f'https://{addr}/exchange'

    subscription_id, watermark = account.calendar.subscribe_to_push(
        callback_url=callback_url
    )
    print(f"{subscription_id=}")
    print(f"{watermark=}")

   account.calendar.unsubscribe(subscription_id)

def main():
    """Entry point"""
    account = connect(**ACCOUNT_CFG.get('exchange', {}))
    subscription(account)

if __name__ == "__main__":
    main()
```

Request XML: 
``` xml
<?xml version='1.0' encoding='utf-8'?>
<s:Envelope
    xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"
    xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages"
    xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">
  <s:Header>
    <t:RequestServerVersion Version="Exchange2016"/>
    <t:TimeZoneContext>
      <t:TimeZoneDefinition Id="FLE Standard Time"/>
    </t:TimeZoneContext>
  </s:Header>
  <s:Body>
    <m:Subscribe>
      <m:PushSubscriptionRequest>
        <t:FolderIds>
          <t:DistinguishedFolderId Id="calendar">
            <t:Mailbox>
              <t:EmailAddress>*********</t:EmailAddress>
              <t:RoutingType>SMTP</t:RoutingType>
              <t:MailboxType>Mailbox</t:MailboxType>
            </t:Mailbox>
          </t:DistinguishedFolderId>
        </t:FolderIds>
        <t:EventTypes>
          <t:EventType>CopiedEvent</t:EventType>
          <t:EventType>CreatedEvent</t:EventType>
          <t:EventType>DeletedEvent</t:EventType>
          <t:EventType>ModifiedEvent</t:EventType>
          <t:EventType>MovedEvent</t:EventType>
          <t:EventType>NewMailEvent</t:EventType>
          <t:EventType>FreeBusyChangedEvent</t:EventType>
        </t:EventTypes>
        <t:StatusFrequency>1</t:StatusFrequency>
        <t:URL>https://************.ngrok.io/exchange</t:URL>
      </m:PushSubscriptionRequest>
    </m:Subscribe>
  </s:Body>
</s:Envelope>
```

Response XML: 
``` xml
<?xml version='1.0' encoding='utf-8'?>
<s:Envelope
    xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
  <s:Header>
    <h:ServerVersionInfo
    xmlns:h="http://schemas.microsoft.com/exchange/services/2006/types" xmlns="http://schemas.microsoft.com/exchange/services/2006/types"
    xmlns:xsd="http://www.w3.org/2001/XMLSchema"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" MajorVersion="15" MinorVersion="1" MajorBuildNumber="2308" MinorBuildNumber="14" Version="V2017_07_11"/>
  </s:Header>
  <s:Body
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:xsd="http://www.w3.org/2001/XMLSchema">
    <m:SubscribeResponse
    xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages"
    xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">
      <m:ResponseMessages>
        <m:SubscribeResponseMessage ResponseClass="Success">
          <m:ResponseCode>NoError</m:ResponseCode>
          <m:SubscriptionId>DwBleDEuaW5kaXYubG9jYWwQAAAABzzY1tyEKUWJQL/VW8xdosd6fdLAVdkIEAAAAIKOTCGdXgZPhNPtJZQTWwI=</m:SubscriptionId>
          <m:Watermark>AQAAAJ+x6mniWEdHkEigKSiKvcmSMmUMAAAAAAA=</m:Watermark>
        </m:SubscribeResponseMessage>
      </m:ResponseMessages>
    </m:SubscribeResponse>
  </s:Body>
</s:Envelope>
```

1. After subscribing, the callback URL isn't called.
2. Unsubscribing throw error: `exchangelib.errors.ErrorInvalidSubscription: Subscription is invalid`.

**Additional context**
python: 3.8.9
exchangelib: 4.2.0
MS Exchange: 2016

**Comment by ecederstrand** [comment]
As for 2., EWS does not support unsubscribing push subscriptions. They are supposedly removed automatically if calling the endpoint fails too many times.

As for 1., what did you try? Are you sure that events happened in the calendar folder that would have caused an event to be triggered?

**Comment by dmitrysync** [comment]
I created and deleted an event.
Using the following code:

``` python
EXAMPLE_EVENT_DATA = {
    "subject": "E1518",
    "body": {
        "content": "Description",
        "contentType": "text"
    },
    "categories": ['Зеленая категория'],
    "isAllDay": False,
    "start": {
        "dateTime": "2021-08-03T18:00:00",
        "timeZone": "Europe/Kiev"
    },
    "end": {
        "dateTime": "2021-08-03T18:15:00",
        "timeZone": "Europe/Kiev"
    },
    "showAs": "busy",
    "sensitivity": "normal",
    "reminderMinutesBeforeStart": 10,
}

def process_date(dateTime: str, timeZone: str):
    datetime = arrow.get(dateTime)
    tz = EWSTimeZone.timezone(timeZone)
    return tz.localize(EWSDateTime(datetime.year, datetime.month, datetime.day, datetime.hour, datetime.minute))


def create_event(account: Account, event_data=EXAMPLE_EVENT_DATA) -> CalendarItem:
    """create event"""
    calendar_item = CalendarItem(
        folder=account.calendar,
        subject=event_data.get('subject'),
        body=event_data.get('body', {}).get('content'),
        categories=event_data.get('categories'),
        is_all_day=event_data.get('isAllDay'),
        start=process_date(**event_data.get('start', {})),
        end=process_date(**event_data.get('end', {})),
        legacy_free_busy_status=event_data.get('showAs', '').capitalize(),
        sensitivity=event_data.get('sensitivity', '').capitalize(),
        reminder_minutes_before_start=event_data.get('reminderMinutesBeforeStart'),
        reminder_is_set=True,
    )

    calendar_item.save()
    return calendar_item
```

``` python
def delete_event(account: Account, event_id: str):
    item = account.calendar.get(id=event_id)
    item.delete()
```

Is this a trigger for call the callback URL?

**Comment by ecederstrand** [comment]
You did subscribe to `CreatedEvent` and `DeletedEvent` on the `calendar` distinguished folder, and you're creating and deleting a calendar item in that folder. So it seems you're doing the right thing. You need to create/delete *after* creating the subscription, of course.

Push subscriptions are not covered in the test suite since creating externally available endpoints in GitHub test runners is not possible, so I can't say for certain that this functionality works in exchangelib. But looking at the XML and assuming that push subscriptions are not broken in Exchange, I would try to look at other parts of your setup. Is the callback URL actually available from the Exchange server? Does it accept all HTTP methods (HEAD, GET, PUT, POST, DELETE etc.)?

**Comment by dmitrysync** [comment]
Yes, I create and delete an event after subscribing.

For testing I use `NgRok`.
This URL is available and accepts all HTTP methods.

But my `ngrok` server isn't receiving any input calls.

**Comment by ecederstrand** [comment]
Ok, then I don't have any other ideas. It may take some time before I have time to play around with push notifications myself.

**Comment by ecederstrand** [comment]
I just tested this using ngrok from my local machine. Here's my test code:

```python
from datetime import datetime, timedelta
from exchangelib import Account, Configuration, Credentials

user, password = 'john@example.com', '********'
email = user
c = Configuration(credentials=Credentials(user, password))
a = Account(email, config=c, autodiscover=True, access_type=DELEGATE)
subscription_id, watermark = a.calendar.subscribe_to_push(callback_url='https://99f04f27b999.ngrok.io')
sleep(60)
start = a.default_timezone.localize(datetime.now())
end = start + timedelta(hours=3)
item = CalendarItem(folder=a.calendar, subject='Hello subscription', start=start, end=end).save()
sleep(60)
item.delete()
```

When I start up the ngrok tunnel (`./ngrok http 80`) and run the above code, I see incoming requests from the Exchange server. When I inspect the requests via http://localhost:4040/inspect/http, I see the following POST request:

```xml
<?xml version="1.0" encoding="utf-8"?>
<soap11:Envelope
    xmlns:soap11="http://schemas.xmlsoap.org/soap/envelope/">
    <soap11:Header>
        <t:RequestServerVersion
            xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages" Version="Exchange2016"
            xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types" />
        </soap11:Header>
        <soap11:Body>
            <m:SendNotification
                xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"
                xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages">
                <m:ResponseMessages>
                    <m:SendNotificationResponseMessage ResponseClass="Success">
                        <m:ResponseCode>NoError</m:ResponseCode>
                        <m:Notification>
                            <t:SubscriptionId>XXXXX=</t:SubscriptionId>
                            <t:PreviousWatermark>AAAAA=</t:PreviousWatermark>
                            <t:MoreEvents>false</t:MoreEvents>
                            <t:StatusEvent>
                                <t:Watermark>AQAAAD9bWvO+LbpDifacly369KukGSQFAAAAAAE=</t:Watermark>
                            </t:StatusEvent>
                        </m:Notification>
                    </m:SendNotificationResponseMessage>
                </m:ResponseMessages>
            </m:SendNotification>
        </soap11:Body>
    </soap11:Envelope>
```

which looks fine.Exchange is telling the HTTP server that it started a subscription towards the URL.  ngrok returns a HTTP 406 on that request, and it's entirely possible that the Exchange server expects a HTTP 200 or something else before it sends notifications for the generated events.

In conclusion, I think creating push subscriptions in exchangelib does indeed work.

**Comment by ecederstrand** [comment]
For anyone else interested in this, I implemented a very simple HTTP server that can receive push notifications from Exchange and parse them. Here's the steps I used to test:

1. start the below Flask app so it's listening on http://localhost:9000
2. Open an ngrok tunnel to the Flask app: `./ngrok http http://localhost:9000`
3. Note down the ngrok URL and paste it in as `callback_url` in the script posted in the comment above
4. Run the above script and see the Flask app printing notifications

```python
from flask import Flask, request
from exchangelib.services import SendNotification

app = Flask(__name__)

@app.route('/', methods=['POST'])
def receive():
    for notification in SendNotification(None).parse(request.data):
        print(notification)
    return "", 201

if __name__ == '__main__':
    app.run(debug=True, host='0.0.0.0', port=9000)
```

**Comment by dmitrysync** [comment]
Hi.

What Exchange server did you use?
Where is your server hosted?

**Comment by ecederstrand** [comment]
My test account is hosted at sfp-net.com. The Exchange build of their server is 15.2.922.13.

**Comment by dmitrysync** [comment]
My server is hosted on OVH.
Subscriptions don't work there.
Please suggest hosting with a working Exchange subscription.

**Comment by ecederstrand** [comment]
I don't have experience with notifications on other hosted Exchange solutions. My test account sits with https://www.skyfillers.com/en.html. You could give that a try.

**Comment by dmitrysync** [comment]
Thanks!
